service/mantle/ups/UpsServices.xml

```diff
@@ -47,4 +47,85 @@
+            </entity-find>
+
+            <!--Call service create#ShipmentRequestMap for the shipment map of request-->
+            <service-call name="mantle.ups.UpsServices.create#ShipmentRequestMap" in-map="context" out-map="shipmentRequest"/>
+            <if condition="!shipmentRequest.shipmentRequestMap">
+                <return type="warning" message="No shipment information is available"/>
+            </if>
+            <set field="shipment" from="[Shipment:shipmentRequest.shipmentRequestMap]"/>
+            <set field="shipmentRequest" from="[ShipmentRequest:shipment]"/>
+
+            <!--Calling generic service for sending request-->
+            <service-call name="mantle.ups.UpsServices.send#UpsRequest" in-map="[shippingGatewayConfigId:shippingGatewayConfigId, path:'ship/v1/shipments', methodType:'POST', requestMap:shipmentRequest]" out-map="shippingLabelResponseOut"/>
+
+            <!-- Get PackageResults from shippingLabelResponseOut to extract image Urls and tracking numbers -->
+            <set field="packageResultsList" from="shippingLabelResponseOut.responseMap.ShipmentResponse.ShipmentResults.PackageResults"/>
+
+            <!-- If shippingLabelResponseOut is empty -->
+            <set field="responseMap" from="shippingLabelResponseOut.responseMap"/>
+            <if condition="!responseMap">
+                <return type="danger" message="Response not received on calling get#ShippingLabel service"></return>
+            </if>
+
+            <!-- Store label image urls and tracking numbers in a list in the same sequence of packages as in request-->
+            <set field="imageUrls" from="[]"/>
+            <set field="trackingNumbers" from="[]"/>
+
+            <if condition="packageResultsList instanceof List">
+                <iterate list="packageResultsList" entry="packageResultInfo">
+                    <set field="graphicImageUrl" from="packageResultInfo.ShippingLabel.GraphicImage"/>
+                    <set field="trackingNumber" from="packageResultInfo.TrackingNumber"/>
+                    <script>imageUrls.add(graphicImageUrl)
+                        trackingNumbers.add(trackingNumber)</script>
+                </iterate>
+                <else>
+                    <!--If packageResult is a Map for single package -->
+                    <set field="graphicImageUrl" from="packageResultsList.ShippingLabel.GraphicImage"/>
+                    <set field="trackingNumber" from="packageResultsList.TrackingNumber"/>
+                    <script>imageUrls.add(graphicImageUrl)
+                        trackingNumbers.add(trackingNumber)</script>
+                </else>
+            </if>
+
+            <!-- Store each Package's trackingCode and labelImage in its record of packageRouteSeg -->
+            <set field="iterator" type="Integer" value="0"/>
+            <iterate list="packageRouteSegList" entry="packageRouteSeg">
```

Comment:
I think you don't need extra variable like iterator to keep track of indexes. You can access the index using packageRouteSeg_index. Please check.